File: bf/bf.py
#!/usr/bin/env python

"""A Brainfuck interpreter"""

__author__="Andrew Pennebaker (andrew.pennebaker@gmail.com)"
__date__="18 Nov 2005 - 6 Jul 2006"
__copyright__="Copyright 2006 Andrew Pennebaker"
__version__="0.5"
__URL__="http://snippets.dzone.com/posts/show/3536"

import sys, os, getopt

tape=[0]*100
address=0

def sublevel(level):
	i=0

	if level.count("[")>level.count("]"):
		raise Exception("Unmatched Bracket")

	while level[0:i+1].count("[")!=level[0:i+1].count("]"):
		i+=1

	return level[1:i+1]

def run(code, position=0):
	global tape
	global address

	while position<len(code):
		cmd=code[position]

		if cmd=="<":
			address-=1
		elif cmd==">":
			address+=1
		elif cmd=="+":
			tape[address]+=1
		elif cmd=="-":
			tape[address]-=1
		elif cmd==".":
			sys.stdout.write(chr(tape[address]))
		elif cmd==",":
			try:
				tape[address]=ord(sys.stdin.read(1))
			except:
				tape[address]=-1
		elif cmd=="[":
			try:
				level=sublevel(code[position:])

				while tape[address]!=0:
					run(level)

				position+=len(level)

			except:
				return position
		elif cmd=="]":
			return position

		position+=1

	return position

def usage():
	print "Usage: %s [options] [<sourcefile>]" % (sys.argv[0])
	print "-h|--help (usage)"

	sys.exit()

def main():
	systemArgs=sys.argv[1:]

	live=False

	optlist, args=[], []

	try:
		optlist, args=getopt.getopt(systemArgs, "h", ["help"])
	except Exception, e:
		usage()

	live=len(args)<1

	for option, value in optlist:
		if option=="-h" or option=="--help":
			usage()

	if live:
		print "--BF Interpreter 0.5--"
		print "  Type exit to exit."

		code=""
		position=0

		line="not exit"
		while line!="exit":
			sys.stdout.write("% ")
			line=sys.stdin.readline().rstrip(os.linesep)
			code+=line

			position=run(code, position)
	else:
		src=args[0]

		srcfile=open(src, "r")
		code="".join(srcfile.readlines())
		srcfile.close()

		if code.count("[")!=code.count("]"):
			raise "Unbalanced brackets"
		else:
			run(code)

if __name__=="__main__":
	try:
		main()
	except KeyboardInterrupt, e:
		pass

File: bf/bf.rb
#!/usr/bin/env ruby

# Author:: Andrew Pennebaker
# Copyright:: Copyright 2009 Andrew Pennebaker
#
# == Synopsis
#
# bf: Brainfuck interpreter
#
# == Usage
#
# bf [options] [script]
#
# --debug, -d:
#    enable debug mode
#
# --help, -h:
#    show help
#
# --version, -v:
#    show version

require 'getoptlong'

BF_APP = 'Brainfuck'
BF_VERSION = '0.0.1'
BF_COPYRIGHT = 'Copyright 2009 YelloSoft'
BF_WELCOME = 'Use \'exit\' or Control-C to quit.\n\n'
BF_INTERACTIVE_EXIT = 'exit'

#
# Brainfuck VM
#
class VM
  attr_accessor :COMMENT, :tape, :address, :code, :pos, :step, :debug

  COMMENT = '#'

  def initialize
    @tape = [0] # Dynamic
    @address = 0
    @code = ''
    @pos = 0
    @step = 0

    @debug = false
  end

  def self.valid?(level)
    level.count('[') == level.count(']')
  end

  # Assumes level is valid code.
  def self.sublevel(level)
    i = 1

    slevel = level[0, i]

    until valid?(slevel)
      i += 1
      slevel = level[0, i]
    end

    slevel[1, slevel.length - 2]
  end

  def run(level, position = 0)
    len = level.length

    while position < len
      @step += 1

      puts self if @debug

      c = level[position, 1]

      case c
      when '+'
        @tape[@address] += 1
      when '-'
        @tape[@address] -= 1
      when '<'
        @address -= 1
      when '>'
        @address += 1

        @tape << 0 if @address == @tape.length
      when '.'
        STDOUT.putc(@tape[@address].chr)
      when ','
        @tape[@address] = STDIN.readchar
      when '['
        slevel = VM.sublevel(level[position, len - position])

        run(slevel) until @tape[@address] == 0

        position += slevel.length + 1
      end

      position += 1
    end

    position
  end

  def evalbf(command)
    result = ''

    command.split('\n').each do |cmd|
      if cmd[0, 1] != COMMENT
        @code += cmd

        if VM.valid?(@code)
          @pos = run(@code, @pos)

          result = @tape[@address].to_s
        end
      end
    end

    result
  end

  def interactive
    welcome

    loop do
      line = gets

      break if line == BF_INTERACTIVE_EXIT

      result = evalbf(line.chomp)

      puts result if result != ''
    end
  end

  def scripted(script)
    lines = ''

    begin
      f = open(script, 'r')

      lines = f.read

      f.close
    rescue
      raise "Error reading file #{script}"
    end

    evalbf(lines)
  end

  def to_s
    "--- Step #{@step} ---
Code:\n#{@code}
Position: #{@pos}
Command: #{@code[@pos, 1]}
Tape: #{ @tape.map { |e| e.to_s }.join(" ") }
Address: #{@address}
Value: #{@tape[@address]}"
  end
end

def version
  puts "#{BF_APP} #{BF_VERSION}"
  exit
end

def welcome
  puts "#{BF_APP} #{BF_VERSION} #{BF_COPYRIGHT}\n#{BF_WELCOME}"
end

def usage
  system("more #{$PROGRAM_NAME}")
  exit
end

def main
  mode = :scripted

  vm = VM.new

  opts = GetoptLong.new(
    ['--debug', '-d', GetoptLong::NO_ARGUMENT],
    ['--help', '-h', GetoptLong::NO_ARGUMENT],
    ['--version', '-v', GetoptLong::NO_ARGUMENT]
  )

  opts.each do |option, value|
    case option
    when '--debug'
      vm.debug = true
    when '--help'
      usgae
    when '--version'
      version
    end
  end

  mode = :interactive if ARGV.length < 1

  case mode
  when :interactive
    vm.interactive
  when :scripted
    vm.scripted(ARGV[0])
  end
end

if $PROGRAM_NAME == __FILE__
  begin
    main
  rescue Interrupt
    nil
  rescue RuntimeError => e
    puts e.message
  end
end
